### PR TITLE
Лабораторная работа 3

### PR DESCRIPTION
**Pull request by kolonist**
Лабораторная работа 3
Основные и дополнительное задание.

### CODE REVIEW

**Comment by medbar** [comment]
В score забыл torch.no_grad

Решение принял (12 + 4 + 4)

Вопрос по работе: 
Чем AdamW отличается от Adam(weight_decay=0.01) ?

**Comment by kolonist** [comment]
> В score забыл torch.no_grad

Да, спасибо. Я уже в других МР-ах прочитал, но т.к. уже отправил (и, главное, обучил), не стал исправлять.

> Чем AdamW отличается от Adam(weight_decay=0.01) ?

В оптимизаторе `Adam` применяется не совсем честный `weight decay`, у него `weight_decay` — это на самом деле L2 регуляризация (или L2 штраф). `weight_decay` здесь применяется для модификации градиента.

У `AdamW` разработчики применили уже честный `weight decay`, который применяется в расчетах не так, как в `Adam`. Здесь `weight_decay` уже применяется для модицикации самих параметров.

В формулах это выглядит так:
image
<sub>Источник: https://arxiv.org/abs/1711.05101</sub>

В коде пайторча это так:

**Adam:**
```python
if weight_decay != 0:
    grad = grad.add(param, alpha=weight_decay)
```

**AdamW:**
```python
# Perform stepweight decay
param.mul_(1 - lr * weight_decay)
```

<sub>(а еще пишут, что в реализации Keras отличий нет, но я сам не видел, так что утверждать не могу)</sub>

**Comment by medbar** [comment]
Баллы выставил :+1:
У тебя достаточно баллов для допуска к экзамену. 4ую можно не сдавать. Если хочешь пойти на экзамен, а не получить автомат, конечно.

**Comment by kolonist** [comment]
Спасибо!
